### PR TITLE
OS X path normalization

### PR DESCRIPTION
**Pull request by mhubig**
Quick fix for @edgewood's PR #171 for #143

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
IMHO normalized should be a function, case normalization and encoding correction comes to mind

File: borg/testsuite/helpers.py [hunk 1]
```diff
@@ -104,6 +104,36 @@ def test(self):
...
+    
+    def testInvalidUnicode(self):
+        pattern = str(b'ba\x80', 'latin1')
+        i = IncludePattern(pattern)
+        ni = IncludePattern(pattern, normalize=True)
+
+        assert not i.match("ba/foo")
+        assert i.match(str(b"ba\x80/foo", 'latin1'))
+        assert not ni.match("ba/foo")
+        assert ni.match(str(b"ba\x80/foo", 'latin1'))
```
Comment (ThomasWaldmann) [inline borg/testsuite/helpers.py]: above, you decode the bytestrings "correctly" with latin1 (as all codes are mapped in latin1, every byte sequence is correct). did you check how borg handles this? does it work with bytes? does it try to decode somehow?

File: borg/testsuite/helpers.py [hunk 2]
```diff
@@ -104,6 +104,36 @@ def test(self):
         self.assert_equal(self.evaluate(['/etc/', '/var'], ['dmesg']),
                           ['/etc/passwd', '/etc/hosts', '/var/log/messages', '/var/log/dmesg'])
 
+class OSXPatternNormalizationTestCase(BaseTestCase):
```
Comment (ThomasWaldmann) [inline borg/testsuite/helpers.py]: the tests themselves do not seem OS X specific.

File: borg/helpers.py [hunk 3]
```diff
@@ -220,20 +222,40 @@ def exclude_path(path, patterns):
...
+            self.normalized = False
 
-    def match(self, path):
-        return (path+os.path.sep).startswith(self.pattern)
+        self.pattern = os.path.normpath(pattern).rstrip(os.path.sep)+os.path.sep
 
     def __repr__(self):
-        return '%s(%s)' % (type(self), self.pattern)
+        return '%s(%s)%s' % (type(self), self.pattern,
+                "normalized" if self.normalized else '')
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: as this repr will look like a contructor call, maybe move the normalize into the parens, like a parameter.

File: borg/helpers.py [hunk 4]
```diff
@@ -220,20 +222,40 @@ def exclude_path(path, patterns):
...
-    def __init__(self, pattern):
-        self.pattern = os.path.normpath(pattern).rstrip(os.path.sep)+os.path.sep
+    def __init__(self, pattern, normalize=False):
+        # inner method used to create the class method 'match'
+        def match(path):
+            return (path+os.path.sep).startswith(self.pattern)
+
+        # HFS+ converts paths to a canonical form, so don't require
+        # users to enter an exact match
+        if normalize: # NOTE: could alternately be if sys.platform in ('darwin'):
```
Comment (ThomasWaldmann) [inline borg/helpers.py]: if OS X is our only use case for this (and it looks like it is), yes, do the platform check here.

self.normalize = sys.platform in ('darwin', )
if ...

Comment (mhubig) [inline borg/helpers.py]: ok

File: borg/helpers.py (hunk 3 above)
Comment (mhubig) [inline borg/helpers.py]: If we use `sys.platform in ('darwin', )` we may just remove it here?

File: borg/testsuite/helpers.py (hunk 2 above)
Comment (mhubig) [inline borg/testsuite/helpers.py]: If we use `sys.platform in ('darwin', )`, they are, so we can put them into a `skipUnless` clause ...

````python
@unittest.skipUnless(sys.platform.startswith('darwin'), 'OS X only test')
class NormalizationTestCase(BaseTestCase):
    def testComposedUnicode(self):
        pattern = 'b\N{LATIN SMALL LETTER A WITH ACUTE}'
        ...
````

File: borg/testsuite/helpers.py (hunk 1 above)
Comment (mhubig) [inline borg/testsuite/helpers.py]: Hmm may be @edgewood can shine a light on this test case ...

**Comment by ThomasWaldmann** [comment]
is this superceded by PR #193 ? if so, close this?

**Comment by mhubig** [comment]
Closed in favor of #193.

**Pull request closed without merging**